ocaml/tests/test_repository_helpers.ml

```diff
@@ -0,0 +1,1218 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Test_highlevel
+open Repository_helpers
+
+let fields_of_pkg = Fmt.Dump.([
+    field "name" (fun (r:Pkg.t) -> r.name) string
+  ; field "version" (fun (r:Pkg.t) -> r.version) string
+  ; field "release" (fun (r:Pkg.t) -> r.release) string
+  ; field "arch" (fun (r:Pkg.t) -> r.arch) string
+  ])
+
+module PkgOfFullnameTest = Generic.MakeStateless (struct
+  module Io = struct
+    type input_t = string
+
+    type output_t = (Pkg.t option, exn) result
+
+    let string_of_input_t x = x
+
+    let string_of_output_t =
+      Fmt.(str "%a" Dump.(result ~ok:(option @@ record @@ fields_of_pkg) ~error:exn))
+  end
+
+  let transform input =
+    try Ok (Pkg.of_fullname input)
+    with e -> Error e
+
+  let tests =
+    `QuickAndAutoDocumented
+      [
+        ("libpath-utils-0.2.1-29.el7.x86",
+         Ok None)
+      ; ("libpath-utils.x86_64",
+         Error Api_errors.(Server_error (internal_error,
+                                         [Pkg.error_msg "libpath-utils.x86_64"])))
+      ; ("libpath-utils-0.2.1-29.el7.noarch",
+         Ok (Some Pkg.{name="libpath-utils"; version="0.2.1"; release="29.el7"; arch="noarch"}))
+      ; ("libpath-utils-0.2.1-29.el7.x86_64",
+         Ok (Some Pkg.{name="libpath-utils"; version="0.2.1"; release="29.el7"; arch="x86_64"}))
+      ]
+end)
+
+let fields_of_update = Fmt.Dump.([
+    field "name" (fun (r:Update.t) -> r.name) string
+  ; field "arch" (fun (r:Update.t) -> r.arch) string
+  ; field "old_version" (fun (r:Update.t) -> r.old_version) (option string)
+  ; field "old_release" (fun (r:Update.t) -> r.old_release) (option string)
+  ; field "new_version" (fun (r:Update.t) -> r.new_version) string
+  ; field "new_release" (fun (r:Update.t) -> r.new_release) string
+  ; field "update_id" (fun (r:Update.t) -> r.update_id) (option string)
+  ])
+
+module UpdateOfJsonTest = Generic.MakeStateless (struct
+  module Io = struct
+    type input_t = Yojson.Basic.t
+
+    type output_t = (Update.t, exn) result
+
+    let string_of_input_t j = Yojson.Basic.to_string j
+
+    let string_of_output_t =
+      Fmt.(str "%a" Dump.(result ~ok:(record @@ fields_of_update) ~error:exn))
+  end
+
+  let transform input =
+    try Ok (Update.of_json input)
+    with e -> Error e
+
+  let tests =
+    `QuickAndAutoDocumented
+      [
+        (`Assoc [
```

Comment:
I think we should use a raw json string, rather than this intermediate Yojson type. `{| ... |}` comes in handy here (you don't need to escape `"`)

```
let ok_json_test_cases = [
  {| { "name": "libpath-utils", ... } |}, Update.{ name="libpath-utils", ... },
  ...
]
```